reset.c

```diff
@@ -0,0 +1,888 @@
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE
+
+*/
+
+#include "samtools.h"
+
+#include "htslib/sam.h"
+#include "sam_opts.h"
+#include "htslib/thread_pool.h"
+#include "htslib/khash.h"
+
+KHASH_SET_INIT_INT(aux_set);        //set hold the aux tags to be retained
+KHASH_SET_INIT_STR(pg_set);         //set hold the pg tags to be removed
+
+#define TAGNUM(X) ((X)[0] << 8) | (X)[1]    //to create key for aux tags, like type key in htslib
+
+typedef struct conf_data
+{
+    int keepRGs;                    //RG line handling
+    int keepPGs;                    //PG line handling
+    khash_t(aux_set) *auxtagsin;    //SET that holds the aux tags to be reatined
+    khash_t(pg_set) *pgout;         //SET that holds the pg tags to be removed
+    int tmpfield;                   //tmp field
+} conf_data;
+
+/// usage - print the usage
+/** @param fp pointer to the file / terminal to which usage to be dumped
+returns nothing
+*/
+static void usage(FILE *fp)
+{
+    fprintf(fp,
+            "Usage: samtools reset [options]\n\
+      -i <file path>        input file\n\
+      -o <file path>        output file\n\
+      -k <file path>        aux tags to be retained in csv file, optional\n\
+      -r <file path>        aux tags to be rejected in csv file, optional\n\
+      --reject-PG [<file path>] pg names to be rejected in csv file, optional\n\
+      --no-RG               to have RG lines or not, optional\n");
+
+    sam_global_opt_help(fp, "--.--.--");
+    return;
+}
+
+/// getdefaultauxtags - add default aux tags to the collection
+/** @param config pointer to conf_data which holds the aux_set pointer; if the aux_set exists, clear and create new one
+returns 1 on failure and 0 on success
+*/
+int getdefaultauxtags(conf_data *config)
+{
+    int ret = 0;
+
+    if (!config) {
+        return 1;
+    }
+
+    if (config->auxtagsin) {      //free if already exists
+        kh_destroy(aux_set, config->auxtagsin);
+        config->auxtagsin = NULL;
+    }
+
+    config->auxtagsin = kh_init(aux_set);    //initialize the new one
+
+    //add defaults, they may get overridden by reject option
+    kh_put(aux_set, config->auxtagsin, TAGNUM("BC"), &ret);   //barcode
+    kh_put(aux_set, config->auxtagsin, TAGNUM("CB"), &ret);   //cell identifier
+    //TODO add tags which are kept for backward compatibility?
+    kh_put(aux_set, config->auxtagsin, TAGNUM("LB"), &ret);   //library
+    kh_put(aux_set, config->auxtagsin, TAGNUM("MI"), &ret);   //molecular identifier
+
+    return 0;
+}
+
+/// getcsvauxtags - aux tag parsed from csv input file, validate and add to internal set
+/** @param field - the aux tag parsed
+ *  @param data - pointer to conf_data which holds internal flags and aux_set
+returns -1 on failure and 0 on success
+*/
+int getcsvauxtags(const char *field, void *data)
```

Comment:
Why is data "void *"?  It's called from one function with type conf_data *, which it is cast into anyway.

Reply:
Done